File: include/thruster.py
import numpy as np

# This is the motor class which will create the motors
class Motor:

    # Initialising the motors
    def __init__(self,motor_specs=None):

        if motor_specs is None:
            motor_specs=[0.0]*4

        self.position=np.array(motor_specs[0:3])    # Creates the position array of individual motor
        self.kf=motor_specs[3]                      # Getting the kf value
        self.km=motor_specs[4]                      # Getting the km value
        self.spin=motor_specs[5]                    # +1 for CW and -1 for ACW
        self.max_RPM=motor_specs[6]                 # Setting the max RPM limits

    def create_motor(self,motor_specs):
        self.position=np.array(motor_specs[0:3])
        self.kf=motor_specs[3]
        self.km=motor_specs[4]
        self.spin=motor_specs[5]
        self.max_RPM=motor_specs[6]

    # Running the motor on the RPM
    def run_motor(self,RPM):
        return np.array([0.0,0.0,self.kf*(RPM**2)])

# Initialising the drone by creating all four motors
def initialise_drone(arm_length,kf,km,max_rpm):

    front_motor=Motor([0.0,arm_length,0.0,kf,km,1.0,max_rpm])
    right_motor=Motor([arm_length,0.0,0.0,kf,km,-1.0,max_rpm])
    back_motor=Motor([0.0,-arm_length,0.0,kf,km,1.0,max_rpm])
    left_motor=Motor([-arm_length,0.0,0.0,kf,km,-1.0,max_rpm])

    return [front_motor,right_motor,back_motor,left_motor]

# Calculating the net thrust by providing the RPM and motor data
def calc_thrust(RPMs,motors):
    fRPM,rRPM,bRPM,lRPM=RPMs
    fmotor,rmotor,bmotor,lmotor=motors

    fT=fmotor.run_motor(fRPM)
    rT=rmotor.run_motor(rRPM)
    bT=bmotor.run_motor(bRPM)
    lT=lmotor.run_motor(lRPM)

    return fT+rT+bT+lT

# Calculating the net required torque by providing the RPM and motor data
def calc_torque(RPMs,motors):

    fRPM,rRPM,bRPM,lRPM=RPMs
    fmotor,rmotor,bmotor,lmotor=motors

    fT=fmotor.run_motor(fRPM)
    rT=rmotor.run_motor(rRPM)
    bT=bmotor.run_motor(bRPM)
    lT=lmotor.run_motor(lRPM)

    fTou=np.cross(fmotor.position,fT)   
    rTou=np.cross(rmotor.position,rT)   
    bTou=np.cross(bmotor.position,bT)   
    lTou=np.cross(lmotor.position,lT)

    resfTou=np.array([0.0,0.0,(fmotor.spin*fmotor.km*(fRPM**2))])   
    resrTou=np.array([0.0,0.0,(rmotor.spin*rmotor.km*(rRPM**2))])   
    resbTou=np.array([0.0,0.0,(bmotor.spin*bmotor.km*(bRPM**2))])   
    reslTou=np.array([0.0,0.0,(lmotor.spin*lmotor.km*(lRPM**2))])   

    return fTou+rTou+bTou+lTou+resfTou+resrTou+resbTou+reslTou

File: src/thruster.cpp
#include "thruster.h"

// Constructors

Motor::Motor() : position(0.0,0.0,0.0),kf(0.0),km(0.0),spin(1),maxRPM(0.0)
{}

Motor::Motor(Vector3 position,double kf,double km,int spin,double maxRPM) : position(position),
                                                                            kf(kf),
                                                                            km(km),
                                                                            spin(spin),
                                                                            maxRPM(maxRPM)
{}

// To run the motor

Vector3 Motor::runMotor(double RPM) const{
    return Vector3(0.0,0.0,kf*(RPM*RPM));   // RUns on the RPM provided
}

// Getter functions

Vector3 Motor::getPosition() const{
    return position;
}

double Motor::getKf() const{
    return kf;
}

double Motor::getKm() const{
    return km;
}

int Motor::getSpin() const{
    return spin;
}

double Motor::getMaxRPM() const{
    return maxRPM;
}

// Some functions to run and initialise the motors

// Creating all four motors of the drone and storing them in a MotorSet
MotorSet initialiseDrone(double armLength,double kf, double km, double maxRPM){

    Motor frontMotor(Vector3(0.0,armLength,0.0),kf,km,1,maxRPM);
    Motor rightMotor(Vector3(armLength,0.0,0.0),kf,km,-1,maxRPM);
    Motor backMotor(Vector3(0.0,-armLength,0.0),kf,km,1,maxRPM);
    Motor leftMotor(Vector3(-armLength,0.0,0.0),kf,km,-1,maxRPM);

    return {frontMotor,rightMotor,backMotor,leftMotor};
}

// Calculating the net thrust based upon the RPMs and the motor datas
Vector3 calcThrust(const std::array<double,4>& RPMs,const MotorSet& motors){

    Vector3 fT=motors.frontMotor.runMotor(RPMs[0]);
    Vector3 rT=motors.rightMotor.runMotor(RPMs[1]);
    Vector3 bT=motors.backMotor.runMotor(RPMs[2]);
    Vector3 lT=motors.leftMotor.runMotor(RPMs[3]);

    return fT+rT+bT+lT;
}

// Calculating the net torque based upon the RPMs and the motor datas
Vector3 calcTorque(const std::array<double,4>& RPMs,const MotorSet& motors){

    Vector3 fT=motors.frontMotor.runMotor(RPMs[0]);
    Vector3 rT=motors.rightMotor.runMotor(RPMs[1]);
    Vector3 bT=motors.backMotor.runMotor(RPMs[2]);
    Vector3 lT=motors.leftMotor.runMotor(RPMs[3]);

    Vector3 fTou=motors.frontMotor.getPosition()^fT; 
    Vector3 rTou=motors.rightMotor.getPosition()^rT; 
    Vector3 bTou=motors.backMotor.getPosition()^bT; 
    Vector3 lTou=motors.leftMotor.getPosition()^lT; 

    Vector3 resfTou(0.0,0.0,(motors.frontMotor.getSpin()*motors.frontMotor.getKm()*(RPMs[0]*RPMs[0])));
    Vector3 resrTou(0.0,0.0,(motors.rightMotor.getSpin()*motors.rightMotor.getKm()*(RPMs[1]*RPMs[1])));
    Vector3 resbTou(0.0,0.0,(motors.backMotor.getSpin()*motors.backMotor.getKm()*(RPMs[2]*RPMs[2])));
    Vector3 reslTou(0.0,0.0,(motors.leftMotor.getSpin()*motors.leftMotor.getKm()*(RPMs[3]*RPMs[3])));

    return fTou+rTou+bTou+lTou+resfTou+resrTou+resbTou+reslTou;
}